lib/acts_as_taggable_on_mongoid/taggable/tag_type_definition.rb

```diff
@@ -0,0 +1,188 @@
+# frozen_string_literal: true
+
+module ActsAsTaggableOnMongoid
+  module Taggable
+    # TagTypeDefinition represents the definition for a aingle tag_type in a model.
+    #
+    # The options passed into a tag_type defined through acts_as_taggable* method are stored in the tag definition
+    # which then drives the creation of the relations and methods that are added to the model.
+    #
+    # The TagTypeDefinition mirrors the Configuraiton attributes and defaults any value that isn't passed in
+    # to the value in the Configuration (ActsAsTaggableOnMongoid.configuration)
+    class TagTypeDefinition
+      attr_reader :owner,
+                  :tag_type
+
+      include ActsAsTaggableOnMongoid::Taggable::TagTypeDefinition::Attributes
+      include ActsAsTaggableOnMongoid::Taggable::TagTypeDefinition::Names
+      include ActsAsTaggableOnMongoid::Taggable::TagTypeDefinition::Changeable
+
+      def initialize(owner, tag_type, options = {})
+        options.assert_valid_keys(:parser,
+                                  :preserve_tag_order,
+                                  :cached_in_model,
+                                  :force_lowercase,
+                                  :force_parameterize,
+                                  :remove_unused_tags,
+                                  :tags_table,
+                                  :taggings_table)
+
+        options.each do |key, value|
+          instance_variable_set("@#{key}", value)
+        end
+
+        @owner    = owner
+        @tag_type = tag_type
+      end
+
+      # rubocop:disable Layout/SpaceAroundOperators
+      # :reek:FeatureEnvy
+
+      # I've defined the parser as being required to return an array of strings.
+      # This parse function will take that array and make it a TagList which will then use the tag_definition
+      # to apply the rules to that list (like case sensitivity and parameterization, etc.) to get the final
+      # list.
+      def parse(*tag_list)
```

Comment:
This is a departure from ActsAsTaggableOn.

There, a parser was expected to return a TagList.  I felt that this was an unreasonable expectation of the parser, and really made my life more difficult.

All parsing that isn't only for a spec is done through this function.  The parser just returns a regular every day array, and this function ensures that it is converted to the right object type with the right values.